src/librustc_privacy/lib.rs

```diff
@@ -919,7 +919,32 @@ impl<'a, 'tcx: 'a> SearchInterfaceForPrivateItemsVisitor<'a, 'tcx> {
     }
 
     fn impl_trait_ref(&mut self) -> &mut Self {
-        self.tcx.impl_trait_ref(self.item_def_id).visit_with(self);
+        if let Some(impl_trait_ref) = self.tcx.impl_trait_ref(self.item_def_id) {
+            // Non-local means public (private items can't leave their crate, modulo bugs)
+            if let Some(node_id) = self.tcx.hir.as_local_node_id(impl_trait_ref.def_id) {
+                let item = self.tcx.hir.expect_item(node_id);
+                let vis = ty::Visibility::from_hir(&item.vis, node_id, self.tcx);
+
+                if !vis.is_at_least(self.min_visibility, self.tcx) {
+                    self.min_visibility = vis;
+                }
+                if !vis.is_at_least(self.required_visibility, self.tcx) {
+                    if self.has_pub_restricted || self.has_old_errors {
+                        struct_span_err!(self.tcx.sess, self.span, E0445,
+                                         "private trait `{}` in public interface", impl_trait_ref)
+                            .span_label(self.span, format!(
+                                        "private trait can't be public"))
+                            .emit();
+                    } else {
+                        self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
+                                               node_id,
+                                               self.span,
+                                               format!("private trait `{}` in public \
+                                                        interface (error E0445)", impl_trait_ref));
+                    }
+                }
+            }
+        }
         self
```

Comment:
Thanks for catching this. Unfortunately, still fails. Somehow we're still not looking at the trait_ref properly.